elasticcloud.py

```diff
@@ -0,0 +1,166 @@
+        service_account          = None
+        
+        with open(service_account_key_path) as f:
+            service_account = json.load(f)
+
+        # From service account key
+        project_id            = service_account['project_id']
+        service_account_email = service_account['client_email']
+
+        Driver = get_driver(Provider.GCE)
+        self.gce = Driver(service_account_email, service_account_key_path, datacenter=datacenter, project=project_id)
+
+        # Used to format node names (datetime formatting)
+        self.format = '%m-%d-%Y-%H-%M-%S'
+        
+
+    def get_oldest_node(self):
+        nodes = self.gce.list_nodes()
+        for n in nodes:
+            print(n.name)
+
+    def get_active_node_quantity(self):
+        return len(self.gce.list_nodes())
+
+    def get_node_names(self):
+        nodes = self.gce.list_nodes()
+        names = []
+        for n in nodes:
+            names.append(n.name)
+        return names
+
+    def get_node_ips(self):
+        nodes = self.gce.list_nodes()
+        ips = []
+        for n in nodes:
+            ips.append(n.public_ips)
+        return ips
+
+    def get_ips(self):
+        return self.get_node_ips()
+
+    def expand(self):
+        if self.get_active_node_quantity() < self.max_nodes:
+            now = datetime.now()
+            node_name = 'gpu-' + now.strftime(self.format)
+            new_node = self.gce.create_node(name=node_name, size=self.size, image=self.image)
+            """
+            new_node = self.gce.create_node(name=node_name,
+                                            size=self.size, 
+                                            image=self.image, 
+                                            ex_accelerator_count=1,
+                                            ex_accelerator_type='nvidia-tesla-p100',
+                                            )
+            """
+
+            self.gce.wait_until_running([new_node])
+            return "New node running at " + new_node.public_ips[0] + " with name " + node_name
+        else:
+            return "Already " + str(self.get_active_node_quantity()) + " nodes running. (max)"
+            
+    def shrink(self):
+        if self.get_active_node_quantity() > self.min_nodes:
+            # Find oldest node, Destroy it
+            # use strptime to parse node names and compare
+            nodes = self.gce.list_nodes()
+            oldest_date = datetime.strptime(nodes[0].name[4:], self.format)
+            oldest_node = nodes[0]
+
+            for node in nodes:
+                date = datetime.strptime(node.name[4:], self.format)
+                if date < oldest_date:
+                   oldest_node = node 
+
+            self.gce.destroy_node(oldest_node)
+            return "Destroyed node: " + oldest_node.name + "."
+        else:
+            return "Only " + self.min_nodes + " nodes running. (min)"
+
+    
+
+
+class ElasticCloud(GCEMixin):
```

Comment:
What's this going to look like when we load in multiple adapters? Below you use `GCEMixin()` even though it's inherited -- I think we should remove this inheritance. 


### Something like this seems to make sense:



```python
ACTION_SHRINK = 'shrink'
ACTION_EXPAND = 'expand'
ACTION_DO_NOTHING = 'do_nothing'

class ElasticCloudAdapter:
    """This class is so we know what an adapter should look like, and if we need any common behavior between them, like loading configuration"""
    def __init__(self):
        self.config = self._load_configuration()
    def _load_configuration(self):
        return dictionary_of_configuration_stuff_somehow
    def expand(self):
        raise NotImplemented()
    def shrink(self):
        raise NotImplemented()


class GCSAdapter(ElasticCloudAdapter):
    def _load_configuration(self):
        super()._load_configuration()  # we can use Super!
        # ... additional tweaks to configuration loading, if needed

   # ... actual implementations of expand, shrink, etc. *specific to GCS*


class AzureAdapter(ElasticCloudAdapter):
   # ... actual implementations of expand, shrink, etc. *specific to Azure*



@click.command()
@click. argument('driver', type=click.Choice(['gcs', 'azure']))
def do_load_check_and_execute_if_needed(driver):
    if driver == "gcs":
       adapter = GCSAdapter()
    # ...
    next_action = adapter.get_next_action()
    if next_action == ACTION_SHRINK:
        adapter.shrink()
    if next_action == ACTION_EXPAND:
        adapter.expand()
```

Then our scheduler runs `do_load_check_and_execute_if_needed gcs` every 5 mins or so! I think that seems nice..

What do you think of that structure? Been a long day, apologies if I'm overlooking something!

Reply:
This structure makes much more sense to me now. I'll switch it up to work like this.